### PR TITLE
Add possibility do data define label position by point additionally to X/Y

### PR DESCRIPTION
**Pull request by domi4484**
This will allow to use the move label tool to change the position of a label coming from a point geometry column (It is assumed that the additional geometry has the same CRS that the main one).
It will also be possible do define just one expression for the label position and not two very similar expressions for X and Y.
Activating the Point override expression will deactivate X/Y overrides, and activating X or Y will deactivate Point.

Gui changes:
image

### CODE REVIEW

**Comment by nyalldawson** [comment]
I'd like to review before.merge

**Comment by roya0045** [comment]
Would this be where the values be set when moving labels?

How does it handle the copy/paste symbology? Sometimes the x & Y fields become invalidated and using free placement doesn't work and the values need to be reset.

File: python/gui/auto_generated/qgstextformatwidget.sip.in [hunk 1]
```diff
@@ -145,11 +145,9 @@ Sets the background color for the text preview widget.
 :param color: background color
 %End
 
-    void enableDataDefinedAlignment( bool enable );
+    void updateDataDefinedAlignment();
```
Comment (m-kuhn) [inline python/gui/auto_generated/qgstextformatwidget.sip.in:148]: @nyalldawson what is the current approach with API changes like this (not often used)?
Can they be documented in api_break.dox or do we need to keep a `( bool enable = false)` and document it as unused param to be removed with QGIS 4?

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 2]
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
...
+    {
+      if ( !attributes.at( pointCol ).isNull() )
+      {
+        QVariant pointAsVariant = attributes.at( pointCol );
+        if ( pointAsVariant.canConvert<QgsGeometry>() )
+        {
+          QgsGeometry geometryPoint = pointAsVariant.value<QgsGeometry>();
+          const QgsPoint *point  = qgsgeometry_cast<QgsPoint *>( geometryPoint.constGet() );
+
+          x = point->x();
```
Comment (m-kuhn) [inline src/app/labeling/qgsmaptoollabel.cpp:726]: Can you check the point for nou null here?

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 3]
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
...
+          y = point->y();
+
+          xSuccess = true;
+          ySuccess = true;
+        }
+        else if ( !pointAsVariant.toByteArray().isEmpty() )
```
Comment (m-kuhn) [inline src/app/labeling/qgsmaptoollabel.cpp:732]: Where are bytearray wkb points produced?

Comment (nyalldawson) [reply src/app/labeling/qgsmaptoollabel.cpp:732]: I'm guessing it's to support (Read only) fields which are already created on the database as wkb.

Personally I don't think this a safe approach -- we're feeding any random binary field into the fromWkb method, without any safeguard that it is actually wkb content. 

I think we could drop this handling here and instead just get users to use an expression based property with a `geom_from_wkb( "wkb_field")` expression if they need this functionality.

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 4]
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
...
+
+    if ( mCurrentLabel.settings.dataDefinedProperties().isActive( QgsPalLayerSettings::PositionPoint ) )
+    {
+      if ( !attributes.at( pointCol ).isNull() )
+      {
+        QVariant pointAsVariant = attributes.at( pointCol );
+        if ( pointAsVariant.canConvert<QgsGeometry>() )
+        {
+          QgsGeometry geometryPoint = pointAsVariant.value<QgsGeometry>();
+          const QgsPoint *point  = qgsgeometry_cast<QgsPoint *>( geometryPoint.constGet() );
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:724]: Hmm, this really shouldn't compile at all -- it should be complaining because the const pointer is being cast to a non-const pointer :scream:

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 5]
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
...
+
+          x = point->x();
+          y = point->y();
+
+          xSuccess = true;
+          ySuccess = true;
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:730]:
```suggestion
         const  QgsGeometry geometry = pointAsVariant.value<QgsGeometry>();
         if ( const QgsPoint *point  = ( geometry.constGet() ? qgsgeometry_cast<QgsPoint *>( geometry.constGet()->simplifiedTypeRef() ) : nullptr ) )
         {
            x = point->x();
            y = point->y();

            xSuccess = true;
            ySuccess = true;
          }
```

File: python/gui/auto_generated/qgstextformatwidget.sip.in (hunk 1 above)
Comment (nyalldawson) [reply python/gui/auto_generated/qgstextformatwidget.sip.in:148]: I'd suggest deprecating the old enableDataDefinedAlignment method and making it a no-op. 

(And then hiding the new updateDataDefinedAlignment() method from sip, as neither should be exposed)

File: src/ui/qgstextformatwidgetbase.ui [hunk 6]
```diff
@@ -5558,6 +5558,33 @@ font-style: italic;</string>
                              </property>
                             </widget>
                            </item>
+                           <item>
+                            <spacer name="horizontalSpacer_6">
```
Comment (nyalldawson) [inline src/ui/qgstextformatwidgetbase.ui:5562]: Looking at the screenshot this layout has resulted in a very wide horizontal minimum size on the widget. Can you instead move the new Point label and button to a new row below the x/y buttons?

File: src/core/labeling/qgspallabeling.h [hunk 7]
```diff
@@ -321,6 +321,7 @@ class CORE_EXPORT QgsPalLayerSettings
       // (data defined only)
       PositionX = 9, //!< X-coordinate data defined label position
       PositionY = 10, //!< Y-coordinate data defined label position
+      PositionPoint = 79, //!< Point-coordinate data defined label position
```
Comment (nyalldawson) [inline src/core/labeling/qgspallabeling.h:324]:
```suggestion
      PositionPoint = 114, //!< Point-coordinate data defined label position
```

I vaguely recall that an entry was removed from this enum in 3.0, which would account for the gap. If that's so then we should avoid reusing the possibly conflicting value

File: src/core/labeling/qgspallabeling.cpp [hunk 8]
```diff
@@ -244,6 +245,7 @@ void QgsPalLayerSettings::initPropertyDefinitions()
     { QgsPalLayerSettings::PolygonLabelOutside, QgsPropertyDefinition( "PolygonLabelOutside", QgsPropertyDefinition::DataTypeString, QObject::tr( "Label outside polygons" ),  QObject::tr( "string " ) + "[<b>yes</b> (allow placing outside)|<b>no</b> (never place outside)|<b>force</b> (always place outside)]", origin ) },
     { QgsPalLayerSettings::PositionX, QgsPropertyDefinition( "PositionX", QObject::tr( "Position (X)" ), QgsPropertyDefinition::Double, origin ) },
     { QgsPalLayerSettings::PositionY, QgsPropertyDefinition( "PositionY", QObject::tr( "Position (Y)" ), QgsPropertyDefinition::Double, origin ) },
+    { QgsPalLayerSettings::PositionPoint, QgsPropertyDefinition( "PositionPoint", QObject::tr( "Position (point)" ), QgsPropertyDefinition::Point, origin ) },
```
Comment (nyalldawson) [inline src/core/labeling/qgspallabeling.cpp:248]:
```suggestion
    { QgsPalLayerSettings::PositionPoint, QgsPropertyDefinition( "PositionPoint", QgsPropertyDefinition::DataTypeString, QObject::tr( "Position (point)" ), QObject::tr( "string " ) + "A point geometry", origin ), origin ) },
```

File: src/core/labeling/qgspallabeling.cpp [hunk 9]
```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
...
+  bool hasDataDefinedPosition = false;
   {
-    exprVal = mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() );
-    if ( !exprVal.isNull() )
+    bool ddPosition = false;
+
+    if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionX )
+         && mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionY )
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() ).isNull()
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() ).isNull() )
```
Comment (nyalldawson) [inline src/core/labeling/qgspallabeling.cpp:2383]: We are double-evaluating these two properties -- can you store the value from the first evaluation to avoid this?

File: src/core/labeling/qgspallabeling.cpp [hunk 10]
```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
...
+      ddPosition = true;
 
-      if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionY ) )
+      QVariant pointAsVariant = mDataDefinedProperties.value( QgsPalLayerSettings::PositionPoint, context.expressionContext() );
+      QgsPoint point;
+      if ( pointAsVariant.canConvert<QgsReferencedGeometry>() )
       {
-        exprVal = mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() );
-        if ( !exprVal.isNull() )
+        point = QgsPoint( pointAsVariant.value<QgsReferencedGeometry>().asPoint() );
```
Comment (nyalldawson) [inline src/core/labeling/qgspallabeling.cpp:2402]: We probably **should** be transforming the point here, not ignoring the point's CRS entirely :wink:

File: src/core/labeling/qgspallabeling.cpp [hunk 11]
```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
...
+      }
+      else if ( pointAsVariant.canConvert<QgsGeometry>() )
+      {
+        point = QgsPoint( pointAsVariant.value<QgsGeometry>().asPoint() );
+      }
+      else if ( !pointAsVariant.toString().isEmpty() )
+      {
+        point.fromWkt( pointAsVariant.toString() );
+      }
```
Comment (nyalldawson) [inline src/core/labeling/qgspallabeling.cpp:2411]: I don't think this logic is correct. I think if a user has a field with WKT, they should use an expression with the geom_from_wkt() function instead (which would mean that the result is always a QgsGeometry)

File: src/core/labeling/qgspallabeling.cpp [hunk 12]
```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
...
+        }
+
+        xPos += xdiff;
+        yPos += ydiff;
+      }
+      else
+      {
+        anchorPosition = QgsPointXY( xPos, yPos );
+
+        // only rotate non-pinned OverPoint placements until other placements are supported in pal::Feature
```
Comment (nyalldawson) [inline src/core/labeling/qgspallabeling.cpp:2524]: this is old --didn't we remove that constraint recently? I suspect there's a merge mistake here which has reintroduced the old code.

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 13]
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
   return true;
 }
 
+bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess, double &y, bool &ySuccess, int &xCol, int &yCol, int &pointCol ) const
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:681]: Can you extend the existing currentLabelDataDefinedPosition method instead of reimplementing it entirely? This is a LOT of duplicate code

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 14]
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
...
+        x = attributes.at( xCol ).toDouble( &xSuccess );
+      if ( !attributes.at( yCol ).isNull() )
+        y = attributes.at( yCol ).toDouble( &ySuccess );
+    }
+  }
+
+  return true;
+}
+
+bool QgsMapToolLabel::layerIsRotatable( QgsVectorLayer *vlayer, int &rotationCol ) const
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:759]: Again, can't we adapt the existing method instead of the duplicate code?

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 15]
```diff
@@ -859,6 +979,34 @@ bool QgsMapToolLabel::isPinned()
   return rc;
 }
 
+bool QgsMapToolLabel::labelMoveable( QgsVectorLayer *vlayer, const QgsPalLayerSettings &settings, int &xCol, int &yCol, int &pointCol ) const
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:982]: More duplicate code

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 16]
```diff
@@ -939,7 +1087,7 @@ bool QgsMapToolLabel::createAuxiliaryFields( LabelDetails &details, QgsPalIndexe
     {
       index = vlayer->fields().lookupField( prop.field() );
     }
-    else
+    else if ( prop.propertyType() != QgsProperty::ExpressionBasedProperty )
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:1090]: This is handled in QgsAuxiliaryLayer::createProperty -- there's logic in there which upgrades an expression based property to allow for combined handling of expressions AND manually placed labels.

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 17]
```diff
@@ -990,7 +1138,7 @@ bool QgsMapToolLabel::createAuxiliaryFields( LabelDetails &details, QgsDiagramIn
...
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:1141]:
```suggestion
    else
```
see above for reason

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 18]
```diff
@@ -1040,7 +1188,7 @@ bool QgsMapToolLabel::createAuxiliaryFields( QgsCalloutPosition &details, QgsCal
...
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:1191]:
```suggestion
    else
```
see above for reason

File: src/app/labeling/qgsmaptoolmovelabel.cpp [hunk 19]
```diff
@@ -221,34 +244,35 @@ void QgsMapToolMoveLabel::cadCanvasPressEvent( QgsMapMouseEvent *e )
         yCol = indexes[ QgsDiagramLayerSettings::PositionY ];
       }
 
-      if ( xCol >= 0 && yCol >= 0 )
+      bool usesAuxFields = vlayer->fields().fieldOrigin( xCol ) == QgsFields::OriginJoin
+                           && vlayer->fields().fieldOrigin( yCol ) == QgsFields::OriginJoin;
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoolmovelabel.cpp:248]: Removing the check is unsafe -- what if xCol or yCol is -1 ?

**Review by nyalldawson — CHANGES_REQUESTED** [review]

File: src/core/labeling/qgspallabeling.cpp (hunk 10 above)
Comment (domi4484) [reply src/core/labeling/qgspallabeling.cpp:2402]: I have introduced a warning if the point crs is not equal to the layer crs. Theoretically it is possible to have a different crs for the additional geometry but i can't see practical cases where it is the case.

**Comment by domi4484** [comment]
I'am finally back to this, thank you very much @nyalldawson for your review last time  🙏

**Comment by nyalldawson** [comment]
@domi4484 
> I will make the old function call the new one. I can't extend it because otherwise will break the api (reference parameters can't have default values).

Those classes are in "app", which isn't constrained by the stable api promise. Feel free to break and adapt them as required! :+1:

**Comment by domi4484** [comment]
> Those classes are in "app", which isn't constrained by the stable api promise. Feel free to break and adapt them as required! +1

Good to know, so I cleaned up a little further.
Not sure if it was a good idea but I also removed the function `QgsMapToolLabel::layerCanPin` which wasn't used anywhere.

**Comment by m-kuhn** [comment]
@nyalldawson does the "changes requested" still apply here?

**Comment by m-kuhn** [comment]
Gentle ping, @nyalldawson

**Comment by nyalldawson** [comment]
@m-kuhn I'll look again when I'm back from leave in another week 👍

File: src/core/labeling/qgspallabeling.cpp (hunk 8 above)
Comment (nyalldawson) [reply src/core/labeling/qgspallabeling.cpp:248]: Because the value **can't** be specified as "[1,2]" -- but I was wrong in suggesting "string" here, it should refer to "geometry". I.e. "Expected format: point geometry value" or similar

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 20]
```diff
@@ -661,18 +662,37 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
...
+    if ( mCurrentLabel.settings.dataDefinedProperties().isActive( QgsPalLayerSettings::PositionPoint ) )
+    {
+      if ( pointCol >= 0
+           && !attributes.at( pointCol ).isNull() )
+      {
+        QVariant pointAsVariant = attributes.at( pointCol );
+        if ( pointAsVariant.canConvert<QgsGeometry>() )
+        {
+          const  QgsGeometry geometry = pointAsVariant.value<QgsGeometry>();
+          if ( const QgsPoint *point  = ( geometry.constGet() ? qgsgeometry_cast<QgsPoint *>( geometry.constGet()->simplifiedTypeRef() ) : nullptr ) )
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:678]:
```suggestion
          if ( const QgsPoint *point  = ( geometry.constGet() ? qgsgeometry_cast<const QgsPoint *>( geometry.constGet()->simplifiedTypeRef() ) : nullptr ) )
```

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 21]
```diff
@@ -848,17 +859,45 @@ bool QgsMapToolLabel::isPinned()
...
+{
+  xCol = -1;
+  yCol = -1;
+  pointCol = -1;
+
+  if ( settings.dataDefinedProperties().isActive( QgsPalLayerSettings::PositionPoint ) )
+  {
+    QString pointColName = dataDefinedColumnName( QgsPalLayerSettings::PositionPoint, settings, vlayer );
+    pointCol = vlayer->fields().lookupField( pointColName );
+    if ( pointCol > 0 )
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:883]:
```suggestion
    if ( pointCol >= 0 )
```

File: src/app/labeling/qgsmaptoollabel.cpp [hunk 22]
```diff
@@ -848,17 +859,45 @@ bool QgsMapToolLabel::isPinned()
...
+  }
+
+  if ( settings.dataDefinedProperties().isActive( QgsPalLayerSettings::PositionX )
+       && settings.dataDefinedProperties().isActive( QgsPalLayerSettings::PositionY ) )
+  {
+    QString xColName = dataDefinedColumnName( QgsPalLayerSettings::PositionX, settings, vlayer );
+    QString yColName = dataDefinedColumnName( QgsPalLayerSettings::PositionY, settings, vlayer );
+    xCol = vlayer->fields().lookupField( xColName );
+    yCol = vlayer->fields().lookupField( yColName );
+    if ( xCol > 0 || yCol > 0 )
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoollabel.cpp:894]:
```suggestion
    if ( xCol >= 0 || yCol >= 0 )
```

File: src/app/labeling/qgsmaptoolmovelabel.cpp [hunk 23]
```diff
@@ -182,16 +182,18 @@ void QgsMapToolMoveLabel::cadCanvasPressEvent( QgsMapMouseEvent *e )
...
 
-      int xCol = -1, yCol = -1;
+      int xCol = -1, yCol = -1, pointCol = -1;
 
-      if ( !mCurrentLabel.pos.isDiagram && !labelMoveable( vlayer, mCurrentLabel.settings, xCol, yCol ) )
+      if ( !mCurrentLabel.pos.isDiagram && !labelMoveable( vlayer, mCurrentLabel.settings, xCol, yCol, pointCol ) )
       {
-        QgsPalIndexes indexes;
+        if ( mCurrentLabel.settings.dataDefinedProperties().isActive( QgsPalLayerSettings::PositionPoint ) )
+          mCurrentLabel.settings.dataDefinedProperties().property( QgsPalLayerSettings::PositionPoint ).setActive( false );
```
Comment (nyalldawson) [inline src/app/labeling/qgsmaptoolmovelabel.cpp:190]: why this change?

**Comment by nyalldawson** [comment]
Getting very close now!

File: src/app/labeling/qgsmaptoolmovelabel.cpp (hunk 23 above)
Comment (domi4484) [reply src/app/labeling/qgsmaptoolmovelabel.cpp:190]: When using the auxiliary layer we store the values as X and Y coordinates as it is now

Comment (nyalldawson) [reply src/app/labeling/qgsmaptoolmovelabel.cpp:190]: So is this situation even possible to hit? If so, I don't think we should just silently throw away the user's data defined configuration for PositionPoint. Better to show a message bar warning instead that the label cannot be moved because of .... and abort early here.

Comment (domi4484) [reply src/app/labeling/qgsmaptoolmovelabel.cpp:190]: Yes it is possible when the point position is defined as a read only expression (not pointing to a writable geometry column).
Ok it make sense to change it like you described.

File: src/app/labeling/qgsmaptoolmovelabel.cpp [hunk 24]
```diff
@@ -221,34 +223,31 @@ void QgsMapToolMoveLabel::cadCanvasPressEvent( QgsMapMouseEvent *e )
         yCol = indexes[ QgsDiagramLayerSettings::PositionY ];
       }
 
-      if ( xCol >= 0 && yCol >= 0 )
```
[the remark this answers is not in the document]
Comment (domi4484) [reply src/app/labeling/qgsmaptoolmovelabel.cpp]: Check is back again

**Comment by domi4484** [comment]
Thanks again @nyalldawson , I addressed all the remaining points

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Can you fix the merge conflicts and then we can merge...?

**Comment by m-kuhn** [comment]
:tada:
